### ISSUE TITLE
Explicit ABCs inheriting from Protocols should not require all abstract methods to be implemented/redeclared.

### ISSUE THREAD

**Issue by DylanLukes** [microsoft/pyright#6586]
**Describe the bug**

Per PEP 544 [here](https://peps.python.org/pep-0544/#merging-and-extending-protocols):

> Subclassing a protocol class would not turn the subclass into a protocol unless it also has typing.Protocol as an explicit base class. **Without this base, the class is “downgraded” to a regular ABC that cannot be used with structural subtyping.**

It is understandable that Pyright would emit an error whenever a plain class inherits from a `Protocol` but does not implement all required methods. However, it is less convincing, especially in light of the above language in the PEP, that an `ABC` inheriting from `Protocol` should also have this requirement.

It doesn't seem good to follow the PEP 544 language to its most liberal conclusion (any class that incompletely implements a protocol should be considered an ABC automatically). However, in the event that a class is marked explicitly as `ABC`, it seems reasonable?

**Notably, the error below _can be avoided_ by re-declaring the abstract methods.** This feels very odd to do, and is extremely unwieldy for e.g. a mixin for a `Protocol` with many methods. It would also imply that an added method on the `Protocol` should require _every mixin for that Protocol to be updated with no change to behavior_, which is much less than ideal.

(This follows up on https://github.com/microsoft/pyright/issues/3730)

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
class Foobar(Protocol):
    @abstractmethod
    def foo(self):
        ...

    @abstractmethod
    @classmethod
    def bar(cls):
        ...


class FooMixin(Foobar, ABC):
    def foo(self):
        print("foo")


class BarMixin(Foobar, ABC):
    @classmethod
    def bar(cls):
        print("bar")
```

This produces:

```sh
  .../scratch.py:16:7 - error: Class derives from one or more protocol classes but does not implement all required members
    Member "bar" is declared in protocol class "Foobar" (reportGeneralTypeIssues)
  .../scratch.py:21:7 - error: Class derives from one or more protocol classes but does not implement all required members
    Member "foo" is declared in protocol class "Foobar" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations 
```

The errors can be avoided by re-declaring the abstract methods:

```py
class FooMixin(Foobar, ABC):
    def foo(self):
        print("foo")

    @abstractmethod
    @classmethod
    def bar(cls):
        ...


class BarMixin(Foobar, ABC):
    @abstractmethod
    def foo(self):
        ...

    @classmethod
    def bar(cls):
        print("bar")
        
# works as expected now
class FooBarImpl(FooMixin, BarMixin, Foobar):
    ...
```

**VS Code extension or command-line**
Running on the command-line, using the newest version (v1.1.338).

**Comment by DylanLukes** [comment]
One further wrinkle: the above workaround cannot work when the `Protocol` member in question is not a method. If it's a property, you're plain out of luck:

```py
class Proto(Protocol):
    x: int

    @abstractmethod
    def foo(self):
        ...


class Mixin(Proto, ABC):
    def foo(self):
        print("foo")


class Concrete(Mixin, Proto):
    x: int = 10

    def foo(self):
        print("bar")
```

```sh
scratch.py:13:7 - error: Class derives from one or more protocol classes but does not implement all required members
    Member "x" is declared in protocol class "Proto" (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
I think it's reasonable to exempt a class from this check if it inherits from a protocol _and_ also from `abc.ABC`.

I've modified the logic accordingly, and this will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.339, which I just published. It will also be included in a future release of pylance.
